GCEWindowsAgent/addresses.go

```diff
@@ -105,101 +254,58 @@ func (a *addressMgr) disabled() (disabled bool) {
 		disabled = *newMetadata.Project.Attributes.DisableAddressManager
 		return disabled
 	}
-	return addressDisabled
-}
-
-func compareIPs(regFwdIPs, mdFwdIPs, cfgIPs []string) (toAdd []string, toRm []string) {
-	for _, mdIP := range mdFwdIPs {
-		if !containsString(mdIP, cfgIPs) {
-			toAdd = append(toAdd, mdIP)
-		}
-	}
-
-	for _, cfgIP := range cfgIPs {
-		if containsString(cfgIP, regFwdIPs) && !containsString(cfgIP, mdFwdIPs) {
-			toRm = append(toRm, cfgIP)
-		}
+	disabled, err = config.Section("Daemons").Key("network_daemon").Bool()
+	if err == nil {
+		return !disabled
 	}
-
-	return
+	return addressDisabled
 }
 
-var badMAC []string
-
 func (a *addressMgr) set() error {
-	ifs, err := net.Interfaces()
-	if err != nil {
-		return err
+	if runtime.GOOS == "windows" {
+		a.applyWSFCFilter()
 	}
 
-	a.applyWSFCFilter()
-
 	for _, ni := range newMetadata.Instance.NetworkInterfaces {
-		mac, err := net.ParseMAC(ni.Mac)
+		iface, err := getInterfaceByMAC(ni.Mac)
 		if err != nil {
 			if !containsString(ni.Mac, badMAC) {
-				logger.Errorf(err.Error())
+				logger.Errorf("Error getting interface: %s", err)
 				badMAC = append(badMAC, ni.Mac)
 			}
 			continue
 		}
 
-		regFwdIPs, err := readRegMultiString(addressKey, mac.String())
-		if err != nil && err != errRegNotExist {
-			logger.Errorf(err.Error())
-			continue
-		} else if err != nil && err == errRegNotExist {
-			// The old agent stored MAC addresses without the ':',
-			// check for those and clean them up.
-			oldName := strings.Replace(mac.String(), ":", "", -1)
-			regFwdIPs, err = readRegMultiString(addressKey, oldName)
-			if err == nil {
-				// Ignore error here as this is just cleanup.
-				deleteRegKey(addressKey, oldName)
-			} else {
-				regFwdIPs = nil
-			}
+		wantIPs := append(ni.ForwardedIps, ni.TargetInstanceIps...)
+		if runtime.GOOS != "windows" {
+			// IP Aliases are not supported on windows.
+			wantIPs = append(wantIPs, ni.IPAliases...)
 		}
 
-		var iface net.Interface
-		for _, i := range ifs {
-			if i.HardwareAddr.String() == mac.String() {
-				iface = i
+		var forwardedIPs []string
+		if runtime.GOOS == "windows" {
+			forwardedIPs, err = getForwardsFromRegistry(ni.Mac)
```

Comment:
This change cause issues as you are not looking at configured IP's, the configured IP's are not persistent after a reboot so we need to compare all three soruces.

Reply:
i think you mean that the registry might have an address which isn't configured. but a reboot would be an apply case (present in metadata, missing in configured) and we only need the registry information for removal cases to determine which addresses have been added by the agent.